### PR TITLE
Add php; Dir Restracture; makefiles

### PR DESCRIPTION
**Pull request by ofir-petrushka**
Add php 5.6 and composer build.
Add makefiles that help local work, see README.
Move build -> base/build

NOTE: this will break dockerhub build for "10m-build"

Please comment, this is a new way of doing some stuff that we actually didn't have an answer to before, how to work locally, will need to go and rewrite old stuff a little, but this seems like progress.

@oryband @doodyparizada

### CODE REVIEW

**Comment by avivl** [comment]
@rounds/server  please take a look

File: base/build/makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+IMAGE_NAME='rounds/10m-build'
+
+all:
+
+build:
+        docker build --tag "$(IMAGE_NAME)" .
```
Comment (oryband) [inline base/build/makefile]: i think 2 spaces or 4 is enough for indent

File: base/build/makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+build_parent:
```
Comment (oryband) [inline base/build/makefile]: can't use dashes instead of underscores? also are these the naming conventions in makefile? i don't know

File: base/build/php/5.6/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,67 @@
+FROM rounds/10m-build
+
+# persistent / runtime deps
+RUN apt-get update && apt-get install -y ca-certificates curl librecode0 libsqlite3-0 libxml2 --no-install-recommends && rm -r /var/lib/apt/lists/*
```
Comment (oryband) [inline base/build/php/5.6/Dockerfile]: break lines

File: base/build/php/5.6/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+RUN mkdir -p $PHP_INI_DIR/conf.d
+
+##<autogenerated>##
+##</autogenerated>##
+
+ENV GPG_KEYS 0BD78B5F97500D450838F95DFE857D9A90D90EC1 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
+	done
```
Comment (oryband) [inline base/build/php/5.6/Dockerfile]:
```dockerfile
RUN set -xe && \
	for key in $GPG_KEYS; do \
		gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
	done
```

File: base/build/php/5.6/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+RUN buildDeps=" \
+		$PHP_EXTRA_BUILD_DEPS \
+		libcurl4-openssl-dev \
+		libreadline6-dev \
+		librecode-dev \
+		libsqlite3-dev \
+		libssl-dev \
+		libxml2-dev \
+		xz-utils \
+	" \
```
Comment (oryband) [inline base/build/php/5.6/Dockerfile]: indentation is off

File: base/build/php/5.6/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+RUN apt-get update && apt-get install -y ca-certificates curl librecode0 libsqlite3-0 libxml2 --no-install-recommends && rm -r /var/lib/apt/lists/*
+
+# phpize deps
+RUN apt-get update && apt-get install -y autoconf file g++ gcc libc-dev make pkg-config re2c --no-install-recommends && rm -r /var/lib/apt/lists/*
+
+ENV PHP_INI_DIR /usr/local/etc/php
+RUN mkdir -p $PHP_INI_DIR/conf.d
+
+##<autogenerated>##
+##</autogenerated>##
```
Comment (oryband) [inline base/build/php/5.6/Dockerfile]: what does this mean? more verbose

File: base/build/php/5.6/composer/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+# https://getcomposer.org/download/
+RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
+
+# install mercurial for hg command
+RUN apt-get update && \ 
+    apt-get install -y mercurial && \
+    apt-get install -y php5-mcrypt php5-mysql && \
+    apt-get clean && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+RUN echo extension=`dpkg -L php5-mcrypt |grep mcrypt.so` > /usr/local/etc/php/conf.d/mcrypt.ini
```
Comment (oryband) [inline base/build/php/5.6/composer/Dockerfile]: spaces between pipes

Comment (oryband) [inline base/build/php/5.6/composer/Dockerfile]: i recommend to use `$(code code code)` instead of backticks
but you'll need to override `sh` to use `bash` for this i think
```dockerfile
RUN echo extension=$(dpkg -L php5-mcrypt | grep mcrypt.so) > /usr/local/etc/php/conf.d/mcrypt.ini
```

File: base/build/php/5.6/composer/makefile [hunk 8]
```diff
@@ -0,0 +1,14 @@
+IMAGE_NAME=rounds/10m-php-composer:5.6
+
+all: 
+
+build:
+	docker build --tag "$(IMAGE_NAME)" .
+
+build_parent:
+	cd .. && $(MAKE) build_cascade
```
Comment (oryband) [inline base/build/php/5.6/composer/makefile]: no `cd` back into dir? are you sure this is how it's done? @doodyparizada ?

**Comment by oryband** [comment]
@ofir-petrushka LGTM
thanks for PRing and adding content! waiting for the fixes

File: base/README.md [hunk 9]
```diff
@@ -5,3 +5,14 @@ Base 10M docker
...
+Way to use locally
+==================
+Some of the images now have makefiles that allow the fallowing: (example usage: make build)
+
+build 		- Build docker image with tagging the right name like it has been published - Know what this means before using!
+build_cascade	- Will cascade the builds from base until your current image locally, see notes on build.
+clean 		- Remove local image by current directory (forced) - this allows you to clean a local build so you will grub the offical ones out of the public repos next time around.
+clean_cascade 	- Removes all images cascading from here up the tree - this allows you to clean all local builds so you will grub the offical ones out of the public repos next time around.
+run		- Run the current image in bash - very usfull for testing.
+update 		- On images that are reusing from public ones. This will update the local image to match the public one with the needed changes.
```
Comment (oryband) [inline base/README.md]: bad indent

**Comment by oryband** [comment]
@ofir-petrushka so i'll need to update the dockerhub base image path after pring this, right?

File: base/README.md [hunk 10]
```diff
@@ -5,3 +5,14 @@ Base 10M docker
 This is the master branch of the tree, all other images inherit from it either directly or by cascade inheritance.
 
 Any change here should cause a cascade rebuild of the entire tree.
+
+Way to use locally
+==================
+Some of the images now have makefiles that allow the fallowing: (example usage: make build)
```
Comment (doodyparizada) [inline base/README.md]: `following` or is it an Australian accent?

File: base/build/makefile [hunk 11]
```diff
@@ -0,0 +1,22 @@
+IMAGE_NAME='rounds/10m-build'
```
Comment (doodyparizada) [inline base/build/makefile]: usually the `M` in makefile is uppercase

Comment (avivl) [inline base/build/makefile]: 3.2 What Name to Give Your Makefile

By default, when make looks for the makefile, it tries the following names, in order: GNUmakefile, makefile and Makefile.

File: base/build/php/5.6/composer/makefile (hunk 8 above)
Comment (doodyparizada) [inline base/build/php/5.6/composer/makefile]: apparently doesn't matter.
for example the following Makefile:
```Makefile
blah:
	cd ~/Workspace; echo `pwd`;
	echo `pwd`
```

runs like so:
```bash
$> make
cd ~/Workspace; echo `pwd`;
/Users/doody/Workspace
echo `pwd`
/tmp
```

File: base/build/php/5.6/composer/makefile [hunk 12]
```diff
@@ -0,0 +1,14 @@
+IMAGE_NAME=rounds/10m-php-composer:5.6
+
+all:
```
Comment (doodyparizada) [inline base/build/php/5.6/composer/makefile]: this is the default and it doesn't do anything.
maybe you can echo the README usage.

File: base/build/php/5.6/docker-php-ext-enable [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+if [ "${#modules[@]}" -eq 0 ]; then
+	usage >&2
+	exit 1
+fi
+
+for module in "${modules[@]}"; do
+	if nm -g "$module" | grep -q ' zend_extension_entry$'; then
+		# https://wiki.php.net/internals/extensions#loading_zend_extensions
+		line="zend_extension=$(readlink -f "$module")"
```
Comment (doodyparizada) [inline base/build/php/5.6/docker-php-ext-enable]: maybe you want the `zend` parts to be private/ change the name?

File: base/build/php/5.6/composer/makefile (hunk 8 above)
Comment (doodyparizada) [inline base/build/php/5.6/composer/makefile]: also, instead of using `cd ..` you can use `make -C ..`

File: base/build/makefile (hunk 1 above)
Comment (ofir-petrushka) [inline base/build/makefile]: Makefile is tab only, can't do spaces, won't run. (stupid but that's make)

File: base/README.md (hunk 10 above)
Comment (ofir-petrushka) [inline base/README.md]: Done.
Awesome gif!

File: base/README.md (hunk 9 above)
Comment (ofir-petrushka) [inline base/README.md]: Done

File: base/build/makefile (hunk 11 above)
Comment (ofir-petrushka) [inline base/build/makefile]: As Aviv mentions it works, but I'll take the upper case.
Done.

File: base/build/makefile (hunk 2 above)
Comment (ofir-petrushka) [inline base/build/makefile]: Done.

File: base/build/php/5.6/Dockerfile (hunk 3 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/Dockerfile]: Won't touch this file, added notes on top.

File: base/build/php/5.6/Dockerfile (hunk 6 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/Dockerfile]: Won't touch this file, added notes on top.

File: base/build/php/5.6/Dockerfile (hunk 4 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/Dockerfile]: Won't touch this file, added notes on top.

File: base/build/php/5.6/Dockerfile (hunk 5 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/Dockerfile]: Won't touch this file, added notes on top.

File: base/build/php/5.6/composer/Dockerfile (hunk 7 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/composer/Dockerfile]: Fixed space.

Comment (ofir-petrushka) [inline base/build/php/5.6/composer/Dockerfile]: changed backticks to $(code) here, not sure why due.

File: base/build/php/5.6/composer/makefile (hunk 12 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/composer/makefile]: Great idea!
done.

File: base/build/php/5.6/composer/makefile (hunk 8 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/composer/makefile]: Tried the 
```make -C ..```
Got errors, left it for another time...

File: base/build/php/5.6/docker-php-ext-enable (hunk 13 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/docker-php-ext-enable]: It's 'zend' the php company not 'zendesk'.
Added notes on the file that it's external...

File: base/build/php/5.6/composer/makefile (hunk 8 above)
Comment (ofir-petrushka) [inline base/build/php/5.6/composer/makefile]: Interesting one Doddy, I'll use. (due I like generic known commands more then redone features)
Done.

File: base/build/php/5.6/Dockerfile (hunk 5 above)
Comment (oryband) [inline base/build/php/5.6/Dockerfile]: hehe i get it

Comment (oryband) [inline base/build/php/5.6/Dockerfile]: - i don't think it's a good convention to have an unready `Dockerfile` hanging around. if this file is a template which is used for generating a "real" Dockerfile, then I suggest you make this explicit and rename these files to have an `.tmpl` extension or something similar.
- in addition i guess you should use some templating tool instead of `sed` etc. like dockerize or even a simple bash tool, as long as it will allow you to generate a new "real" file from the "template" file mentioned above.
- also if you're using `peru` then you have to state it's a dependency somehow and have `make dep` or something install the dependencies

sorry i didn't notice this before, this is kind of a big pr

**Comment by oryband** [comment]
hey @ofir-petrushka 
want to take another go at this pr?
if you think it's ready to be merged and no more work is needed, i'll be happy to merge it. waiting for your response

also, please note we plan on migrating the images to use the trusted builds such as https://hub.docker.com/_/buildpack-deps/ , https://hub.docker.com/_/python/
maybe you should do the same with this one + php

**Comment by ofir-petrushka** [comment]
@oryband 
Added some of the stuff you requested:
* `make dep` to install peru and warn on some requirement.
* remote stuff is now pulled into directories called 'cached' to seperate them from normal code. Still have warnings at the start about not changing them outside of `make update`.
Others:
* More code re-use in the makefiles.
* Travis tests parallel for each container - will fail until you make dependent ones publish under rounds.
* `make private` - White labeling!

I think you missed the meaning of some items, I am trying to reuse public offical images, with some small rewriting like in "base/build/php/5.6/".
Basically I still want to have the same baseline so I have simple commands to re-tree them.
The "cache" dirs don't have to exist in the repos, they are kind of backup for emergency to still allow building even if something catastrophic happens upstream. (deleted git repos?)

Some of the ideas of the big single tree:
* Size / speed of pulling due to better reuse.
* Common items expected. Less work figuring stuff out that is very likely.
* Solving issues in common areas.
* Control for security and other reasons.

If you wanna have a talk about some of the stuff outside of the PR, we can find the time.

**Comment by oryband** [comment]
@ofir-petrushka thank you for the time you took to do all this. After looking at the changes and thinking about it for quite a long time, I've decided to reject this PR.

Let me explain:

1. I liked the travis image build tests you've written, and I'll copy and add them in a separate PR, but without relying on any Makefiles or external scripts.
1. The other ideas you use in this PR are ok (e.g. Makefiles, cascading, etc.), but they deviate from what we need here at Rounds and create a complexity we want to avoid:
 1. I just prefer to `docker run ...` instead of `make ...`
 1. Even though the Makefile inclusion chain (base-->include-->include common) is ok as an idea, it is overly complex for our daily use.
 1. We stopped using peru and instead just use git sparse checkout. I wrote a 10 line sh script that accepts repo dependencies as argument and clones specific sub-directories wherever you need. This drops the python3 dependency for peru, and in addition it has some troublesome bugs and design I encountered in the past and I'd like to avoid: buildinspace/peru#136 buildinspace/peru#128
1. I'd like to keep the directory structure flat e.g. no `base/build`.
1. We're starting to move away from using a single "base" image for everything, and build specific ones for each new purpose. See `sensu/` as an example. Conceptually using a base image would've proved useful, but in practice most of the stuff that's installed is unused. Sometimes there are simple common enough things to add as a base e.g. `pip` for `python-alpine/` base image, but I couldn't find anything common for a generic `base/` image.
1. I also plan to split this repo into different ones, one for each build e.g. `rounds/sensu-docker`, `rounds/elk-docker`, etc. . This probably means I'll be merging your PHP build only to delete it along with this repo later. Originally I thought a single repo for all public images might be a good idea, but not anymore. Mainly due to the reason that it's hard for people to find and notice what they're looking for according to the "10m" title. If someone is looking for an `elk-docker` image, it will be more helpful to have a repo with just that and not "bloated" with other dockerfiles. If the person is interested in other public dockerfiles we're using, then he can just check or organization page.

Still, thank you very much for the effort, and sorry it took me so much time to realize all this. You don't have to agree with everything, and can continue your plans with this repo by forking.

**Comment by oryband** [comment]
@ofir-petrushka following on what I said, see #27 . Every issue has to mention which image it's discussing about. I don't like cramping unspecific images together anymore, and it feels like a bad practice. What I see in GitHub nowadays is people opening a repo per-image. This is easier to manage and more importantly, much easier for people to search, find, and contribute.
